src/main/filesystem/watcher.js

```diff
@@ -96,61 +126,113 @@ class Watcher {
-      .on('change', pathname => change(win, pathname, type, this._preferences.getPreferedEOL()))
+      .on('add', pathname => {
+        if (!this._shouldIgnoreEvent(win.id, pathname, type)) {
+          add(win, pathname, type, this._preferences.getPreferedEOL())
+        }
+      })
+      .on('change', pathname => {
+        if (!this._shouldIgnoreEvent(win.id, pathname, type)) {
+          change(win, pathname, type, this._preferences.getPreferedEOL())
+        }
+      })
       .on('unlink', pathname => unlink(win, pathname, type))
-      .on('addDir', pathname => addDir(win, pathname))
-      .on('unlinkDir', pathname => unlinkDir(win, pathname))
-      .on('raw', (event, path, details) => {
+      .on('addDir', pathname => addDir(win, pathname, type))
+      .on('unlinkDir', pathname => unlinkDir(win, pathname, type))
+      .on('raw', (event, subpath, details) => {
         if (global.MARKTEXT_DEBUG_VERBOSE >= 3) {
-          console.log(event, path, details)
+          console.log('watcher: ', event, subpath, details)
         }
 
-        // rename syscall on Linux (chokidar#591)
+        // Fix atomic rename on Linux (chokidar#591).
+        // TODO: This should also apply to macOS.
+        // TODO: Do we need to rewatch when the watched directory was renamed?
         if (isLinux && type === 'file' && event === 'rename') {
-          const { watchedPath } = details
-          // Use the same watcher and re-watch the file.
-          watcher.unwatch(watchedPath)
-          watcher.add(watchedPath)
+          clearTimeout(renameTimeout)
+          renameTimeout = setTimeout(async () => {
+            renameTimeout = null
+            if (disposed) return
+
+            const fileExists = await exists(watchPath)
+            if (fileExists) {
+              // File still exists but we need to rewatch the file because the inode has changed.
+              watcher.unwatch(watchPath)
+              watcher.add(watchPath)
+            }
+          }, 150)
         }
       })
       .on('error', error => {
-        const msg = `Watcher error: ${error}`
-        console.log(msg)
-        log.error(msg)
+        // Check if too many file descriptors are opened and notify the user about this issue.
+        if (error.code === 'ENOSPC') {
+          if (!enospcReached) {
+            enospcReached = true
+            log.warn('inotify limit reached: Too many file descriptors are opened.')
+
+            win.webContents.send('AGANI::show-notification', {
+              title: 'inotify limit reached',
+              type: 'warning',
+              message: 'Cannot watch all files and file changes because too many file descriptors are opened.'
+            })
+          }
+        } else {
+          log.error(error)
+        }
       })
 
+    const closeFn = () => {
+      disposed = true
+      if (this.watchers[id]) {
+        delete this.watchers[id]
+      }
+      clearTimeout(renameTimeout)
```

Comment:
Check `renameTimeout` is not null, before clearTimeout.

Reply:
See above `clearTimeout` on `null` is valid.